lib/rspec-puppet/adapters.rb

```diff
@@ -237,7 +237,14 @@ def setup_puppet(example_group)
         case RSpec.configuration.facter_implementation.to_sym
         when :rspec
           if supports_facter_runtime?
-            Puppet.runtime[:facter] = proc { RSpec::Puppet::FacterTestImpl.new }
+            # We always prefer an already initialized FacterImpl instead of
+            # creating new instances of FacterTestImpl
+            Puppet.runtime[:facter] = if defined? FacterImpl
+                                        FacterImpl
+                                      else
+                                        proc { RSpec::Puppet::FacterTestImpl.new }
+                                      end
+
             set_facter_impl(Puppet.runtime[:facter])
```

Comment:
The order here could lead to some odd thing. `set_facter_impl` makes sure `FacterImpl` is defined so I'd expect it the other way around. First call `set_facter_impl` and then use `if defined?` In fact, after calling `set_facter_impl` you can simply rely on `FacterImpl`, right?

So
```ruby
set_facter_impl(RSpec::Puppet::FacterTestImpl.new)
Puppet.runtime[:facter] = FacterImpl
```

Reply:
nice catch, thanks!